kernel/src/scan/mod.rs

```diff
@@ -264,15 +267,17 @@ impl Scan {
 
         let global_state = Arc::new(self.global_scan_state());
         let scan_data = self.scan_data(engine)?;
-        let mut scan_files = vec![];
-        for data in scan_data {
-            let (data, vec) = data?;
-            scan_files =
-                state::visit_scan_files(data.as_ref(), &vec, scan_files, scan_data_callback)?;
-        }
-        scan_files
-            .into_iter()
-            .map(|scan_file| -> DeltaResult<_> {
+        let scan_files_iter = scan_data
+            .map_ok(|(data, vec)| {
+                let scan_files = vec![];
+                state::visit_scan_files(data.as_ref(), &vec, scan_files, scan_data_callback)
+                    .map(IntoIterator::into_iter)
+            })
+            .flatten_ok()
+            .flatten_ok();
```

Comment:
agree with @nicklan that there's probably some room for improvement on flatten_ok()'s. how about we just make a follow-up issue to track it and not block this PR it?

Reply:
https://github.com/delta-incubator/delta-kernel-rs/issues/361